Remove BinderOpts from BoundNodes

333fred opened the pull request:
We had a few locations in the bound nodes were we were hanging onto Binders for later use by IOperation. However, this is now largely unnecessary as, after the recent work to bind parameters up front, the only thing the binder was being used for was for access to a Compilation object, which the IOperation factory already had. Fixes https://github.com/dotnet/roslyn/issues/49218.

Commit 1 is the simple stuff, removing `BinderOpt` from methods where it was simple to do so. The commit is largely mechanical, and should likely be reviewed on its own to make reviewing the next commits easier.
Commit 2 removes the binder from using statements and using local declarations.
Commit 3 removes it from foreach loops.
Commit 4 fixes up an IOperation bug around mishandling expanded for GetEnumerator calls. This is a port of [#49534 (commit 5)](https://github.com/dotnet/roslyn/pull/49534/commits/ca6943f0a0d9f315e8a6f5bf5e2c1807d4d16005). I still need to write up a couple of VB tests here, but I'll take care of that on Monday.
Commit 5 is a very small refactoring.
Commit 6 fixes a bug with nullable reference analysis of default parameters in extension `GetEnumerator` methods.

Commits 2-6 are likely best reviewed together, as I did a couple of refactorings in 3 and 4 on the original implementation in 2.

333fred commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5851,6 +5843,8 @@ private VariableState GetVariableState(Optional<LocalState> localState)
                 refKinds.AddRange(argumentRefKindsOpt);
             }
 
+            var binder = _binder ?? compilation.GetBinder(callSyntax);
```
@AlekseyTs this is the location I was talking about yesterday where I managed to get the binder usage down to just one call: http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/OverloadResolution/MethodTypeInference.cs,1334 (the rest of the usages in the chain can be replaced by passing the nested thing on the binder it's accessing). Unfortunately, that last call doesn't seem to be easily fixable. The only time `_binder` should be null is when we're binding the body of an unbound lambda, trying to figure out the return type of that lambda. I'm not certain getting the binder from the compilation is the right thing to do here, particularly around speculation and potentially getting the wrong binder for such cases. Do you have any thoughts?

I still have a bit of other work to do on this PR around using/foreach dispose argument binding, so no rush on this question.

AlekseyTs replied:
> The only time _binder should be null is when we're binding the body of an unbound lambda, trying to figure out the return type of that lambda. I'm not certain getting the binder from the compilation is the right thing to do here, particularly around speculation and potentially getting the wrong binder for such cases. Do you have any thoughts?

I believe an UnboundLambda itself holds a reference to a Binder, indirectly. Can we use that binder? 

---
In reply to: [547419319](https://github.com/dotnet/roslyn/pull/50105#discussion_r547419319) [](ancestors = 547419319)

333fred replied:
Looks like this an option I missed, thanks.

---
In reply to: [548001479](https://github.com/dotnet/roslyn/pull/50105#discussion_r548001479) [](ancestors = 548001479,547419319)

AlekseyTs commented:
>                 binder: binder,

It looks like this binder is used here only to see if we are ```InAttributeArgument```. I do not think getting a binder from compilation would give us the right value if we get here when we are "in attribute" argument. If it is even possible to get here for a case like that. Consider simply passing the boolean here instead of the Binder and figuring out is value through other means. 

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5860 in e48bf0e. [](commit_id = e48bf0e7e33fc190b2425e7e1693b09411d2b177, deletion_comment = False)

AlekseyTs commented:
>                 binder: binder,

> Consider simply passing the boolean here instead of the Binder and figuring out the value through other means. 

Perhaps not worth looking at right now if we are unable to get rid of the Binder in the ```MethodTypeInferrer.Infer``` call below and are able to get the binder from lambda.

---
In reply to: [750330609](https://github.com/dotnet/roslyn/pull/50105#issuecomment-750330609) [](ancestors = 750330609)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5860 in e48bf0e. [](commit_id = e48bf0e7e33fc190b2425e7e1693b09411d2b177, deletion_comment = False)

333fred commented:
>                 binder: binder,

I wasn't able to remove the binder usage below, so I'll leave this as is.

---
In reply to: [750343184](https://github.com/dotnet/roslyn/pull/50105#issuecomment-750343184) [](ancestors = 750343184,750330609)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5860 in e48bf0e. [](commit_id = e48bf0e7e33fc190b2425e7e1693b09411d2b177, deletion_comment = False)

333fred commented:
@dotnet/roslyn-compiler this is ready for review.

jcouv commented:
There's a number of test failures. I'll hold off reviewing for now.

333fred commented:
> There's a number of test failures. I'll hold off reviewing for now.

@jcouve it's just one minor assert I forgot to adjust after fixing the IOperation bug. Doesn't affect the majority of the code, will be fixed shortly.

333fred commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -8498,7 +8487,7 @@ protected override void VisitForEachExpression(BoundForEachStatement node)
+            if (node.EnumeratorInfoOpt?.GetEnumeratorInfo.Method is { IsExtensionMethod: true, Parameters: var parameters } enumeratorMethod)
             {
                 // this is case 7
                 var (method, results, _) = VisitArguments(
```
Note to self: On monday, fix the bug 2 lines below and add a test for nullable with params/default parameter extension methods.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -78535,6 +78535,61 @@ static class Extensions
+            comp.VerifyDiagnostics(
+                // (8,27): warning CS8620: Argument of type 'C<string>' cannot be used for parameter 'c' of type 'C<string?>' in 'IEnumerator<int> CExt.GetEnumerator<string?>(C<string?> c, string? t = null)' due to differences in the nullability of reference types.
+                //         foreach (var i in c)
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "c").WithArguments("C<string>", "C<string?>", "c", "IEnumerator<int> CExt.GetEnumerator<string?>(C<string?> c, string? t = null)").WithLocation(8, 27)
```
This is a new warning. Previously, we were not analyzing the default parameter. This matches the behavior you'd get if you did `c.GetEnumerator()` yourself.

333fred commented:
@dotnet/roslyn-compiler this is ready for review.

jcouv commented on src/Compilers/CSharp/Portable/Binder/MethodArgumentInfo.cs:
```diff
@@ -0,0 +1,28 @@
+        BitVector DefaultArguments,
+        bool Expanded)
+    {
+        public static MethodArgumentInfo ParameterlessMethod(MethodSymbol method)
```
>ParameterlessMethod [](start = 41, length = 19)

nit: "CreateParameterlessMethod"?

jcouv commented on src/Compilers/CSharp/Portable/Binder/ForEachEnumeratorInfo.cs:
```diff
@@ -50,48 +45,41 @@ internal sealed class ForEachEnumeratorInfo
             Debug.Assert((object)collectionType != null, "Field 'collectionType' cannot be null");
             Debug.Assert(elementType.HasType, "Field 'elementType' cannot be null");
-            Debug.Assert((object)getEnumeratorMethod != null, "Field 'getEnumeratorMethod' cannot be null");
+            Debug.Assert((object)getEnumeratorInfo != null, "Field 'getEnumeratorMethod' cannot be null");
```
>getEnumeratorMethod [](start = 68, length = 19)

Update message with new name or nameof? (also below)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs:
```diff
@@ -124,13 +125,27 @@ private BoundStatement RewriteEnumeratorForEachStatement(BoundForEachStatement n
+
+            // If the GetEnumerator call is an extension method, then the first argument was left unconverted by initial binding. We want to replace
+            // the first argument with our converted receiver and pass null as the receiver instead.
+            if (getEnumeratorInfo.Method.IsExtensionMethod)
```
Are we fixing the behavior or merely moving existing logic?

333fred replied:
Mainly moving, but this logic wasn't necessary in the base because we did the binding of the arguments right here, rather than in initial binding.

---
In reply to: [549493533](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493533) [](ancestors = 549493533)

jcouv commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -1145,6 +1175,23 @@ private MethodSymbol PerformForEachPatternOverloadResolution(TypeSymbol patternT
+                {
+                    var argsToParams = overloadResolutionResult.ValidResult.Result.ArgsToParamsOpt;
+                    var expanded = overloadResolutionResult.ValidResult.Result.Kind == MemberResolutionKind.ApplicableInExpandedForm;
+                    BindDefaultArguments(
```
Did we already have tests for MoveNextAsync/DisposeAsync instance or extension methods with default parameter values?

333fred replied:
Neither can be extensions. Looks like we have a test for `MoveNextAsync`: `CodeGenAsyncForEachTests.TestMoveNextAsync_WithOptionalParameter`, but not for dispose. I'll add one.

---
In reply to: [549493544](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493544) [](ancestors = 549493544)

jcouv commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs:
```diff
@@ -6320,6 +6320,716 @@ Element Values(0)
+        }
+
+        [Fact]
+        public void ForEach_ExtensionGetEnumeratorDefaultParam()
```
Consider using a default value other than zero and also testing execution if not already covered elsewhere

333fred replied:
See `CodeGenForEachTests.TestGetEnumeratorPatternViaExtensionWithOptionalParameter` and `CodeGenAsyncForEachTests.TestGetAsyncEnumeratorPatternViaExtensionWithOptionalParameter`.

---
In reply to: [549493550](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493550) [](ancestors = 549493550)

jcouv commented on src/Compilers/CSharp/Portable/Binder/MethodArgumentInfo.cs:
```diff
@@ -0,0 +1,28 @@
+    /// Information about the arguments of a call that can turned into a BoundCall later without recalculating
+    /// default arguments.
+    /// </summary>
+    internal sealed record MethodArgumentInfo(
```
nit: I don't think we need a record here. We don't need value equality semantics. 
On the other hand, I understand it's a nice shorthand to avoid typing constructor... 
Since this is our first record usage, I'm curious what others think.

333fred replied:
I did think about that. It seemed like a nice place to introduce them to the compiler, but I won't cry if you all ask me to just use a regular class here instead.

---
In reply to: [549493565](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493565) [](ancestors = 549493565)

AlekseyTs replied:
> I don't think we need a record here. We don't need value equality semantics. 

Feels like an overkill to me too.

---
In reply to: [549502386](https://github.com/dotnet/roslyn/pull/50105#discussion_r549502386) [](ancestors = 549502386,549493565)

jcouv commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -1547,5 +1622,40 @@ internal override SyntaxNode ScopeDesignator
+        private MethodArgumentInfo BindDefaultArguments(MethodSymbol method, BoundExpression extensionReceiverOpt, bool allowExtensionMethods, bool expanded, SyntaxNode syntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(!method.IsExtensionMethod || allowExtensionMethods);
+            Debug.Assert(extensionReceiverOpt != null || !allowExtensionMethods);
```
Can this assertion be strengthened? `(extensionReceiverOpt != null) == allowExtensionMethods`

jcouv commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -1547,5 +1622,40 @@ internal override SyntaxNode ScopeDesignator
+            Debug.Assert(!method.IsExtensionMethod || allowExtensionMethods);
+            Debug.Assert(extensionReceiverOpt != null || !allowExtensionMethods);
+
+            var argsBuilder = ArrayBuilder<BoundExpression>.GetInstance(method.ParameterCount);
```
Can we avoid grabbing a builder when no args? 
Something like `var args = method.IsExtensionMethod ? ImmutableArray.Create(extensionReceiverOpt) : ImmutableArray.Empty;`.

Consider adding assertion on `method.ParameterCount` (zero unless extension, right?)

333fred replied:
> zero unless extension, right?

No, consider `GetAsyncEnumerator(CancellationToken = default)`. We could check the parameter count and bail early for the 0 case though.

---
In reply to: [549493609](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493609) [](ancestors = 549493609)

jcouv commented on src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs:
```diff
@@ -170,49 +170,68 @@ internal static BoundStatement BindUsingStatementOrDeclarationFromParts(SyntaxNo
+                        var argumentsBuilder = ArrayBuilder<BoundExpression>.GetInstance(disposeMethod.ParameterCount);
+                        ImmutableArray<int> argsToParams = default;
+                        bool expanded = disposeMethod.HasParamsParameter();
+                        originalBinder.BindDefaultArguments(
```
Did we already have tests for using declaration where the Dispose/Async method has default parameter values?

333fred replied:
Non-async does not match with params/default. We have `await using ()` tests with a default parameter (`CodeGenAwaitUsingTests.TestPatternBasedDisposal_InstanceMethod_ReturningInt`), but not using declaration tests, so I'll add one.

---
In reply to: [549493661](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493661) [](ancestors = 549493661)

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3518,17 +3517,17 @@ private int GetOrCreatePlaceholderSlot(object identifier, TypeWithAnnotations ty
             Conversions conversions)
         {
-            var walker = new NullableWalker(compilation,
+            var walker = new NullableWalker(binder.Compilation,
```
>binder.Compilation, [](start = 44, length = 19)

Can we remove the compilation from NullableWalker constructor (since we always have binder now)?

333fred replied:
Probably, but I would want to do that in a separate PR. This is already much bigger than I had originally been intending.

---
In reply to: [549493682](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493682) [](ancestors = 549493682)

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -8498,19 +8487,22 @@ protected override void VisitForEachExpression(BoundForEachStatement node)
-                    argsToParamsOpt: default,
-                    defaultArguments: default,
+                    argsToParamsOpt: enumeratorMethodInfo.ArgsToParamsOpt,
+                    defaultArguments: enumeratorMethodInfo.DefaultArguments,
```
Do we need a similar change for default arguments in disposal in using-declaration?

333fred replied:
I don't believe so, because I don't believe there is any way to observe a nullability issue there. We made the decision to not warn at the call site for default parameters that don't fit the type, and there is no way to get generic substitution at the method level because we don't allow extension methods for using statements/declarations. If there's a case you can think of I'm open to it, but we don't even analyze the method parameters at all today for those cases.

---
In reply to: [549493697](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493697) [](ancestors = 549493697)

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -78535,6 +78535,61 @@ static class Extensions
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,27): warning CS8620: Argument of type 'C<string>' cannot be used for parameter 'c' of type 'C<string?>' in 'IEnumerator<int> CExt.GetEnumerator<string?>(C<string?> c, string? t = null)' due to differences in the nullability of reference types.
```
>C<string?> [](start = 114, length = 10)

I'm not following something. Where does the `C<string?>` come from?

333fred replied:
Because the `T t = default`. `string` comes from the first argument, and the second argument changes that to nullable because it must be.

---
In reply to: [549493723](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493723) [](ancestors = 549493723)

jcouv reviewed (commented):
Done with review pass (iteration 6)

jcouv reviewed (approved):
LGTM Thanks (iteration 7)

333fred commented:
@dotnet/roslyn-compiler for a second review.

AlekseyTs commented:
Looking

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                         if (isAsync)
                         {
                             Debug.Assert(enumeratorType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)));
+                            Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1);
```
>Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1); [](start = 28, length = 62)

Why is this a valid assumption to make?

333fred replied:
From our offline discussion, I'll extract the `IEnumerable`/`IAsyncEnumerable` step out into a separate method and inline getting the methods, so it's clear we're dealing with well-known members matched by signature.

---
In reply to: [549813225](https://github.com/dotnet/roslyn/pull/50105#discussion_r549813225) [](ancestors = 549813225)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                         {
                             Debug.Assert(enumeratorType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)));
+                            Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1);
+                            Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional);
```
>Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional); [](start = 28, length = 67)

Why is this a valid assumption to make? Can this be a paramarray?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
+                            Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1);
+                            Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional);
+
+                            // IAsyncEnumerable<T>.GetAsyncEnumerator has a default param, so let's fill it in
```
>// IAsyncEnumerable<T>.GetAsyncEnumerator has a default param, so let's fill it in [](start = 28, length = 82)

How do we know that? How do we know we are dealing with ```IAsyncEnumerable<T>.GetAsyncEnumerator```? How do we know that there is a default parameter to fill in, even if it is a ```IAsyncEnumerable<T>.GetAsyncEnumerator```? Where did we check that? We cannot make an assumption like that solely based on the fact that this is how the method is defined in the standard framework.

333fred replied:
Extracted into a separate method to make it clear we're dealing with WellKnown or Special members.

---
In reply to: [549815596](https://github.com/dotnet/roslyn/pull/50105#discussion_r549815596) [](ancestors = 549815596)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                         }
+                        else
+                        {
+                            // We know that IEnumerable<T>.GetEnumerator has no parameters
```
>// We know that IEnumerable<T>.GetEnumerator has no parameters [](start = 28, length = 62)

Similar questions here. I think that rather than making assumptions like that we should simply have a helper that looks at the signature and does the right thing. based on the method in hands.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -1145,6 +1175,23 @@ private MethodSymbol PerformForEachPatternOverloadResolution(TypeSymbol patternT
                     // We don't want to allow this case.
                     result = null;
```
>result = null; [](start = 20, length = 14)

It feels like this is a premature optimization, we probably should still make sure that we can build default arguments if they are needed. Also, external consumers might still be interested in getting information about this method. Probably should follow up separately.

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/50164](https://github.com/dotnet/roslyn/issues/50164) to follow up.

---
In reply to: [549831554](https://github.com/dotnet/roslyn/pull/50105#discussion_r549831554) [](ancestors = 549831554)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -1547,5 +1622,45 @@ internal override SyntaxNode ScopeDesignator
             }
         }
+
+        private MethodArgumentInfo GetSpecialTypeMemberInfo(SpecialMember member, SyntaxNode syntax, DiagnosticBag diagnostics)
```
>GetSpecialTypeMemberInfo [](start = 35, length = 24)

From this name it is not obvious that it can handle only methods without parameters.

333fred replied:
I'll also add an assert that the result was null or parameterless.

---
In reply to: [549837359](https://github.com/dotnet/roslyn/pull/50105#discussion_r549837359) [](ancestors = 549837359)

AlekseyTs commented:
>                                                     knownToImplementIDisposable: enumeratorInfoOpt.NeedsDisposal && (object)enumeratorInfoOpt.GetEnumeratorMethod != null ?

Is this condition now checked elsewhere?

---
Refers to: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1640 in fe18ef1. [](commit_id = fe18ef1509731b884ca8ed3b76207e49bae9665b, deletion_comment = True)

AlekseyTs commented on src/Dependencies/PooledObjects/ArrayBuilder.cs:
```diff
@@ -486,13 +486,23 @@ public void AddRange(ImmutableArray<T> items, int length)
 
+        public void AddRange(ImmutableArray<T> items, int start, int length)
+        {
+            Debug.Assert(start + length <= items.Length);
```
>length [](start = 33, length = 6)

Negative length Ok?

AlekseyTs replied:
>start [](start = 25, length = 5)

Negative start Ok?

333fred commented:
>                                                     knownToImplementIDisposable: enumeratorInfoOpt.NeedsDisposal && (object)enumeratorInfoOpt.GetEnumeratorMethod != null ?

It can't be null. We assert that when creating a `ForEachEnumeratorInfo`.

---
In reply to: [752233745](https://github.com/dotnet/roslyn/pull/50105#issuecomment-752233745) [](ancestors = 752233745)

---
Refers to: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1640 in fe18ef1. [](commit_id = fe18ef1509731b884ca8ed3b76207e49bae9665b, deletion_comment = True)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs:
```diff
@@ -465,44 +473,34 @@ private BoundBlock MakeExpressionUsingStatement(BoundUsingStatement node, BoundB
 
-            var argsBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+#if DEBUG
+            Debug.Assert(!method.IsExtensionMethod || method.ParameterRefKinds.IsDefaultOrEmpty || method.ParameterRefKinds[0] != RefKind.Ref);
```
>method.ParameterRefKinds.IsDefaultOrEmpty [](start = 54, length = 41)

Is this method able to handle ref/out/in parameters?

333fred replied:
If I added them to `MethodArgumentInfo`, then it probably could. However, that isn't necessary anywhere we want to use this today.

---
In reply to: [549857233](https://github.com/dotnet/roslyn/pull/50105#discussion_r549857233) [](ancestors = 549857233)

333fred replied:
I also am not sure what to make of a call that had no explicit arguments and also had by-ref parameters.

---
In reply to: [549861482](https://github.com/dotnet/roslyn/pull/50105#discussion_r549861482) [](ancestors = 549861482,549857233)

AlekseyTs replied:
> If I added them to MethodArgumentInfo, then it probably could. However, that isn't necessary anywhere we want to use this today.

Perhaps we should assert unconditionally that there are no such parameters.

---
In reply to: [549861770](https://github.com/dotnet/roslyn/pull/50105#discussion_r549861770) [](ancestors = 549861770,549861482,549857233)

333fred replied:
> Perhaps we should assert unconditionally that there are no such parameters.

Actually, I was incorrect: we do support one refkind, `in` on the first parameter. However, passing default as the refkinds parameter is the correct thing to do since we did not have an `in` in source, and it will be handled by `GetEffectiveArgumentRefKinds` during the process of `MakeCall`.

---
In reply to: [549870955](https://github.com/dotnet/roslyn/pull/50105#discussion_r549870955) [](ancestors = 549870955,549861770,549861482,549857233)

333fred replied:
I can tighten the assertion, however.

---
In reply to: [549874598](https://github.com/dotnet/roslyn/pull/50105#discussion_r549874598) [](ancestors = 549874598,549870955,549861770,549861482,549857233)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                         {
                             Debug.Assert(enumeratorType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)));
+                            Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1);
+                            Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional);
```
>Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional); [](start = 28, length = 67)

I think it is possible to create a test scenario when this parameter is not declared as optional. Let's add a unit-test and observe behavior.

333fred replied:
I've created one, we don't handle it correctly today either. Will be fixed in the next commit.

---
In reply to: [549859342](https://github.com/dotnet/roslyn/pull/50105#discussion_r549859342) [](ancestors = 549859342)

AlekseyTs commented:
Done with review pass (commit 7)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [752251211](https://github.com/dotnet/roslyn/pull/50105#issuecomment-752251211) [](ancestors = 752251211)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:
```diff
@@ -4796,6 +4885,51 @@ public async Task<int> DisposeAsync()
+        }
+        System.Console.Write(""Done"");
+    }
+    public Enumerator GetAsyncEnumerator()
```
>() [](start = 40, length = 2)

Can this method have optional parameters? Do we have tests covering this scenario?

333fred replied:
See `GetAsyncEnumerator_WithOptionalParameter` and `GetAsyncEnumerator_WithParams` in this file.

---
In reply to: [550213376](https://github.com/dotnet/roslyn/pull/50105#discussion_r550213376) [](ancestors = 550213376)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:
```diff
@@ -4796,6 +4885,51 @@ public async Task<int> DisposeAsync()
+    }
+    public sealed class Enumerator
+    {
+        public async Task<bool> MoveNextAsync()
```
>() [](start = 45, length = 2)

Can this method have optional parameters? Do we have tests covering this scenario?

333fred replied:
See `TestMoveNextAsync_WithOptionalParameter` and `TestMoveNextAsync_WithParamsParameter` in this file.

---
In reply to: [550213568](https://github.com/dotnet/roslyn/pull/50105#discussion_r550213568) [](ancestors = 550213568)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:
```diff
@@ -4796,6 +4885,51 @@ public async Task<int> DisposeAsync()
+    }
+    public sealed class Enumerator
+    {
+        public async Task<bool> MoveNextAsync()
```
>MoveNextAsync [](start = 32, length = 13)

Can this method be an extension method? Do we have tests covering this scenario?

333fred replied:
No, this cannot be an extension.

---
In reply to: [550214694](https://github.com/dotnet/roslyn/pull/50105#discussion_r550214694) [](ancestors = 550214694)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -891,21 +809,143 @@ EnumeratorResult createPatternBasedEnumeratorResult(ref ForEachEnumeratorInfo.Bu
+                        diagnostics, errorLocationSyntax.Location, isOptional: false);
+
+                    // Well-known members are matched by signature: we shouldn't find it if it doesn't have exactly 1 parameter.
+                    Debug.Assert(getEnumeratorMethod.ParameterCount == 1);
```
>getEnumeratorMethod [](start = 33, length = 19)

Possibly dereferencing a null ref? Is there a test hole?

333fred replied:
Looks like we're missing the standard missing member tests for all of the async tests. I'll add them.

---
In reply to: [550225429](https://github.com/dotnet/roslyn/pull/50105#discussion_r550225429) [](ancestors = 550225429)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -891,21 +809,143 @@ EnumeratorResult createPatternBasedEnumeratorResult(ref ForEachEnumeratorInfo.Bu
             }
+        }
+
+        private EnumeratorResult SatisfiesIEnumerableInterfaces(ref ForEachEnumeratorInfo.Builder builder, ref BoundExpression collectionExpr, bool isAsync, DiagnosticBag diagnostics, BoundExpression unwrappedCollectionExpr, TypeSymbol unwrappedCollectionExprType)
```
> ref BoundExpression collectionExpr [](start = 106, length = 35)

It doesn't look like there is need to pass this parameter by ref. In fact, it doesn't look like there is a need to pass it at all because we assign to it before we use it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -891,21 +809,143 @@ EnumeratorResult createPatternBasedEnumeratorResult(ref ForEachEnumeratorInfo.Bu
+                    Debug.Assert(getEnumeratorMethod.ParameterCount == 1);
+                    if (getEnumeratorMethod?.Parameters[0].IsOptional == false)
+                    {
+                        collectionExpr = unwrappedCollectionExpr;
```
>collectionExpr = unwrappedCollectionExpr; [](start = 24, length = 41)

Is this assignment redundant?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:
```diff
@@ -4237,6 +4237,95 @@ public IAsyncEnumerator<int> GetAsyncEnumerator(System.Threading.CancellationTok
+}
+";
+            var comp = CreateCompilationWithTasksExtensions(source);
+            comp.VerifyDiagnostics(
```
>VerifyDiagnostics [](start = 17, length = 17)

VerifyEmitDiagnostics?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:
```diff
@@ -4237,6 +4237,95 @@ public IAsyncEnumerator<int> GetAsyncEnumerator(System.Threading.CancellationTok
+}
+";
+            var comp = CreateCompilationWithTasksExtensions(source);
+            comp.VerifyDiagnostics(
```
>VerifyDiagnostics [](start = 17, length = 17)

VerifyEmitDiagnostics?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:
```diff
@@ -4237,6 +4237,95 @@ public IAsyncEnumerator<int> GetAsyncEnumerator(System.Threading.CancellationTok
+            comp.VerifyDiagnostics(
+                // (8,33): error CS8411: Asynchronous foreach statement cannot operate on variables of type 'C' because 'C' does not contain a suitable public instance or extension definition for 'GetAsyncEnumerator'
+                //         await foreach (var i in c)
+                Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "c").WithArguments("C", "GetAsyncEnumerator").WithLocation(8, 33)
```
>Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "c").WithArguments("C", "GetAsyncEnumerator").WithLocation(8, 33) [](start = 16, length = 117)

Is this a breaking change? It looks like we simply used default value for the token in this case. In fact, I am surprised to see a difference in handling for the two scenarios (prior to the latest adjustment). It shouldn't matter whether we use the interface directly or use a type implementing the interface. Both implement the interface, an interface "implements" itself.

333fred replied:
So, this is a breaking change, but the code shouldn't have been allowed. The method definition does not have a default parameter, and the spec explicitly calls out that these methods must be able to called with no parameters. I can send an email to compat council if you'd like, but the number of people who would be affected by such a change is very, very small, and they have to be operating in an unsupported scenario by polyfilling these APIs anyway.

---
In reply to: [550235389](https://github.com/dotnet/roslyn/pull/50105#discussion_r550235389) [](ancestors = 550235389)

333fred replied:
It's also worth noting that we regressed this scenario in master: this test case will null ref in the local rewriter in master currently, because we changed the way default argument binding works and this case was missed.

---
In reply to: [550290615](https://github.com/dotnet/roslyn/pull/50105#discussion_r550290615) [](ancestors = 550290615,550235389)

AlekseyTs replied:
> It's also worth noting that we regressed this scenario in master: this test case will null ref in the local rewriter in master currently, because we changed the way default argument binding works and this case was missed.

I think we should restore the original behavior. The purpose of the change to move creation of default values in binding vs. lowering was a refactoring rather than an introduction of a behavior change. We don't know what else might be broken.

---
In reply to: [550291362](https://github.com/dotnet/roslyn/pull/50105#discussion_r550291362) [](ancestors = 550291362,550290615,550235389)

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/50182](https://github.com/dotnet/roslyn/issues/50182) to track actually erroring in this scenario.

---
In reply to: [550331559](https://github.com/dotnet/roslyn/pull/50105#discussion_r550331559) [](ancestors = 550331559,550291362,550290615,550235389)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:
```diff
@@ -4237,6 +4237,95 @@ public IAsyncEnumerator<int> GetAsyncEnumerator(System.Threading.CancellationTok
+        {
+        }
+    }
+    IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(System.Threading.CancellationToken token) => throw null;
```
>IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(System.Threading.CancellationToken token) => throw null; [](start = 4, length = 119)

It looks like how the interface is implemented makes no difference. Consider adding a test covering implicit implementation as well.

333fred replied:
It does make a difference, actually. Implicit implementation will error out between lines 757-762 in ForEachLoopBinder.cs, because the implicit implementation should have matched the pattern-based step.

---
In reply to: [550235898](https://github.com/dotnet/roslyn/pull/50105#discussion_r550235898) [](ancestors = 550235898)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -891,21 +809,143 @@ EnumeratorResult createPatternBasedEnumeratorResult(ref ForEachEnumeratorInfo.Bu
+
+        private EnumeratorResult SatisfiesIEnumerableInterfaces(ref ForEachEnumeratorInfo.Builder builder, ref BoundExpression collectionExpr, bool isAsync, DiagnosticBag diagnostics, BoundExpression unwrappedCollectionExpr, TypeSymbol unwrappedCollectionExprType)
+        {
+            if (!AllInterfacesContainsIEnumerable(ref builder, unwrappedCollectionExprType, isAsync, diagnostics, out bool foundMultipleGenericIEnumerableInterfaces))
```
>AllInterfacesContainsIEnumerable [](start = 17, length = 32)

Shouldn't we match an interface to itself here? Looks like this is a preexisting condition, perhaps it is worth following up later.

AlekseyTs commented:
Done with review pass (commit 10)

AlekseyTs commented:
>         public async System.Threading.Tasks.Task<bool> MoveNextAsync(int ok = 0)

Consider adjusting this test to use non-default value here and printing what we get at runtime.

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:636 in 9c65989. [](commit_id = 9c65989787e3c222fdd54b51817b3f9fc52c1abb, deletion_comment = False)

AlekseyTs commented:
>         public async System.Threading.Tasks.Task<bool> MoveNextAsync(params int[] ok)

Consider adjusting this test to print what we get at runtime, i.e. an instance of an empty array.

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:671 in 9c65989. [](commit_id = 9c65989787e3c222fdd54b51817b3f9fc52c1abb, deletion_comment = False)

333fred commented:
@AlekseyTs addressed feedback with the strategy we discussed offline. This does result in an observable behavior change in one indexed property IOperation test, but I believe the change is in line with the strategy of not representing bad invocations as anything other than an IInvalidOperation.

---
In reply to: [752671310](https://github.com/dotnet/roslyn/pull/50105#issuecomment-752671310) [](ancestors = 752671310)

333fred commented:
>             Debug.Assert(parameter.IsOptional);

The only change I made, other than making this a local function, was to remove this assert and move it into the parent function.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1167 in 9c65989. [](commit_id = 9c65989787e3c222fdd54b51817b3f9fc52c1abb, deletion_comment = True)

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:
```diff
@@ -1389,10 +1306,18 @@ private BoundExpression GetDefaultParameterSpecialNoConversion(SyntaxNode syntax
-                if (!visitedParameters[parameter.Ordinal] && parameter.IsOptional)
+                if (!visitedParameters[parameter.Ordinal])
                 {
+                    // Params array is filled in the local rewriter
```
In the old LocalRewriter code, the params array would have been filled in before default parameters were looked at, so this wasn't necessary there.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:
```diff
@@ -201,8 +201,13 @@ private BoundIndexerAccess BindIndexerDefaultArguments(BoundIndexerAccess indexe
+                BitVector defaultArguments = default;
                 Debug.Assert(parameters.Length == indexerAccess.Indexer.Parameters.Length);
-                BindDefaultArguments(indexerAccess.Syntax, parameters, argumentsBuilder, refKindsBuilderOpt, ref argsToParams, out var defaultArguments, indexerAccess.Expanded, enableCallerInfo: true, diagnostics);
+                if (indexerAccess.OriginalIndexersOpt.IsDefault)
```
>(indexerAccess.OriginalIndexersOpt.IsDefault) [](start = 19, length = 45)

Consider adding a comment that otherwise we are dealing with a failure.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:
```diff
@@ -1389,10 +1306,19 @@ private BoundExpression GetDefaultParameterSpecialNoConversion(SyntaxNode syntax
+                if (!visitedParameters[parameter.Ordinal])
                 {
+                    // Params array is filled in the local rewriter
+                    if (expanded && parameter.Ordinal == parameters.Length - 1)
```
>if (expanded && parameter.Ordinal == parameters.Length - 1) [](start = 20, length = 59)

Alternatively we could adjust the loop condition for the expanded case.

AlekseyTs reviewed (approved):
LGTM (commit 13), assuming CI is passing.
